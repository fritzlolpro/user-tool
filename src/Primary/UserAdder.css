.adding-widget {
  display: inline-block;
  margin: 0 auto;
}
.icon-button {
  margin-top: 48px !important;
  margin-bottom: 64px !important;
}
