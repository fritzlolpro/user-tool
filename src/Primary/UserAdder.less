.adding-widget {
    display: inline-block;
    margin: 0 auto;
}

.icon-button() {
    margin-top: 48px;
    margin-bottom: 64px;
}

.icon-button {
    .icon-button() !important;
}