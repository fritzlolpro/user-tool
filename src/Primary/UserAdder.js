import React, {Component} from 'react';
import ErrorText from '../Output/Errors.js'
import UserAddForm from '../Forms/UserAddForm.js'
import FloatingActionButton from 'material-ui/FloatingActionButton';
import AddIcon from 'material-ui/svg-icons/social/plus-one';
import Snackbar from 'material-ui/Snackbar';
import './UserAdder.css';


class UserAdder extends Component {
    constructor(props) {
        super(props)
        this.state = {
            adding: false,
            error: false,
            newUser: {
                name: {
                    first: null,
                    last: null
                },
                picture: null
            },
            notificationOpen: false,
            notificationMessage: ''
        }
    }

    handleUserListUpdate = (user) => {
        if (!this.state.error) {
            this
                .props
                .handleUserListUpdate(user)
        }
    }

    handleNotificationsClose = () => {
        this.setState({notificationOpen: false});
    }

    inputFieldsEmpty = () => {
        // checks if some fields in user form are empty
        const lastName = !!this.state.newUser.name.last
        const firstName = !!this.state.newUser.name.first
        const picture = !!this.state.newUser.picture
        return !(lastName && firstName && picture)
    }

    callUsersApi = (options, path) => this
        .props
        .apiCall(options, path)

    onChangeField = (e) => {
        const target = e.target
        const value = target.value
        const name = target.name
        const {newUser} = this.state || []

        if (name === 'name.first') {
            newUser.name.first = value
        }

        if (name === 'name.last') {
            newUser.name.last = value
        }

        if (name === 'picture') {
            newUser.picture = value
        }

        this.setState({newUser})

    }

    onAdding = () => {
        this.setState({error: false, adding: true})
    }

    onDismissError = () => {
        this.setState({error: false})
    }

    onCancelAdd = () => {
        this.setState({adding: false})
    }

    handleErrors = (response) => {

        if (!response.ok) {
            this.setState({error: response.statusText, notificationMessage: 'Error occurred'})
            this.setState(prevState => ({notificationOpen: true}))
        } else {
            this.setState({notificationMessage: 'User posted'})
            this.setState(prevState => ({notificationOpen: true}))
        }

        return response;
    }

    emitErrorOrSuccessOnAdd = (error) => {
        console.log(error)
        return (
            'Error occurred'
        )
    }

    postUser = () => {
        const {newUser} = this.state
        const options = {
            method: 'post',
            body: JSON.stringify(newUser)
        }

        this
            .callUsersApi(options)
            .then(response => this.handleErrors(response))
            .then(this.onCancelAdd(), this.handleUserListUpdate(newUser))
            .catch(error => this.setState({error: error.toString()}))
    }

    renderErrorMsgOrButton = (error) => {

        return (error
            ? <ErrorText onClose={this.onDismissError} error={error}>Problems while posting user
                    <br/>
                    Server says: {error}</ErrorText>
            : <FloatingActionButton
                secondary={true}
                className='icon-button'
                onClick={this.onAdding}>
                <AddIcon/>
            </FloatingActionButton>)
    }

    render() {
        // if error => show error;; show button by default, show form if addin' stuff
        const {adding, error, notificationOpen, notificationMessage} = this.state
        return (
            <div className='adding-widget'>
            {adding
            ? <UserAddForm
                    cancel={this.onCancelAdd}
                    post={this.postUser}
                    onChange={this.onChangeField}
                    disablePostButton={this.inputFieldsEmpty}/>
            : <div>
                {this.renderErrorMsgOrButton(error)}
                <Snackbar
                    open={notificationOpen}
                    message={notificationMessage}
                    autoHideDuration={4000}
                    onRequestClose={this.handleNotificationsClose}/>
            </div>}
            </div>
            )
    }
}

export default UserAdder