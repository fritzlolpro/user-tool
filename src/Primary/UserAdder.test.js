import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import UserAdder from './UserAdder';

import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

describe('UserAdder', () => {

    it('should be defined', () => {
        expect(UserAdder).toBeDefined();
    });

    it('renders without crashing'),
    () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <MuiThemeProvider>
            <UserAdder/>
        </MuiThemeProvider>, div);
    }

    test('has a valid snapshot', () => {
        const component = renderer.create(
            <MuiThemeProvider>
                <UserAdder/>
            </MuiThemeProvider>
        )
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
})
