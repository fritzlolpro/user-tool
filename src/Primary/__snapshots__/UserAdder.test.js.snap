// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAdder has a valid snapshot 1`] = `
<div
  className="adding-widget"
>
  <div>
    <div
      className="icon-button"
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0,0,0,0)",
          "backgroundColor": "transparent",
          "borderRadius": "50%",
          "boxShadow": "0 3px 10px rgba(0, 0, 0, 0.16),
                 0 3px 10px rgba(0, 0, 0, 0.23)",
          "boxSizing": "border-box",
          "color": "rgba(0, 0, 0, 0.87)",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "muiPrepared": true,
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        }
      }
    >
      <button
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
            "backgroundColor": "#ff4081",
            "border": 10,
            "borderRadius": "50%",
            "boxSizing": "border-box",
            "cursor": "pointer",
            "display": "inline-block",
            "fontFamily": "Roboto, sans-serif",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "height": 56,
            "margin": 0,
            "muiPrepared": true,
            "outline": "none",
            "overflow": "hidden",
            "padding": 0,
            "position": "relative",
            "textAlign": "center",
            "textDecoration": "none",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "verticalAlign": "bottom",
            "width": 56,
          }
        }
        tabIndex={0}
        type="button"
      >
        <div
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
        >
          <div
            style={
              Object {
                "muiPrepared": true,
                "top": 0,
                "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              }
            }
          >
            <svg
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              style={
                Object {
                  "color": "#ffffff",
                  "display": "inline-block",
                  "fill": "#ffffff",
                  "height": 56,
                  "lineHeight": "56px",
                  "muiPrepared": true,
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                  "userSelect": "none",
                  "width": 24,
                }
              }
              viewBox="0 0 24 24"
            >
              <path
                d="M10 8H8v4H4v2h4v4h2v-4h4v-2h-4zm4.5-1.92V7.9l2.5-.5V18h2V5z"
              />
            </svg>
          </div>
        </div>
      </button>
    </div>
    <div
      open={false}
      style={
        Object {
          "bottom": 0,
          "display": "flex",
          "left": "50%",
          "muiPrepared": true,
          "position": "fixed",
          "transform": "translate(-50%, 48px)",
          "transition": "transform 400ms cubic-bezier(0.23, 1, 0.32, 1) 0ms, visibility 400ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "visibility": "hidden",
          "zIndex": 2900,
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.87)",
            "borderRadius": 2,
            "flexGrow": 0,
            "fontFamily": "Roboto, sans-serif",
            "height": 48,
            "lineHeight": "48px",
            "maxWidth": 568,
            "minWidth": 288,
            "muiPrepared": true,
            "padding": "0 24px",
            "width": "auto",
          }
        }
        width={3}
      >
        <div
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 14,
              "muiPrepared": true,
              "opacity": 0,
              "transition": "opacity 400ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            }
          }
        >
          <span>
            
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
